src/xdr/Stellar-contract.x

```diff
@@ -0,0 +1,199 @@
+ *    0x_NNNN_NNNN_NNNN_NNN5  - static: void, true, false, ...
+ *    0x_IIII_IIII_TTTT_TTT7  - object: 32-bit index I, 28-bit type code T
+ *    0x_NNNN_NNNN_NNNN_NNN9  - symbol: up to 10 6-bit identifier characters
+ *    0x_NNNN_NNNN_NNNN_NNNb  - bitset: up to 60 bits
+ *    0x_CCCC_CCCC_TTTT_TTTd  - status: 32-bit code C, 28-bit type code T
+ *    0x_NNNN_NNNN_NNNN_NNNf  - reserved
+ *
+ * Up here in XDR we have variable-length tagged disjoint unions but no
+ * bit-level packing, so we can be more explicit in their structure, at the
+ * cost of spending more than 64 bits to encode many cases, and also having
+ * to convert. It's a little non-obvious at the XDR level why there's a
+ * split between SCVal and SCObject given that they are both immutable types
+ * with value semantics; but the split reflects the split that happens in
+ * the implementation, and marks a place where different implementations of
+ * immutability (CoW, structural sharing, etc.) will likely occur.
+ */
+
+// A symbol is up to 10 chars drawn from [a-zA-Z0-9_], which can be packed
+// into 60 bits with a 6-bit-per-character code, usable as a small key type
+// to specify function, argument, tx-local environment and map entries
+// efficiently.
+typedef string SCSymbol<10>;
+
+enum SCValType
+{
+    SCV_U63 = 0,
+    SCV_U32 = 1,
+    SCV_I32 = 2,
+    SCV_STATIC = 3,
+    SCV_OBJECT = 4,
+    SCV_SYMBOL = 5,
+    SCV_BITSET = 6,
+    SCV_STATUS = 7
+};
+
+% struct SCObject;
+
+enum SCStatic
+{
+    SCS_VOID = 0,
+    SCS_TRUE = 1,
+    SCS_FALSE = 2
+};
+
+enum SCStatusType
+{
+    SST_OK = 0,
+    SST_UNKNOWN_ERROR = 1
+    // TODO: add more
+};
+
+union SCStatus switch (SCStatusType type)
+{
+case SST_OK:
+    void;
+case SST_UNKNOWN_ERROR:
+    uint32 unknownCode;
+};
+
+union SCVal switch (SCValType type)
+{
+case SCV_U63:
+    uint64 u63;
+case SCV_U32:
+    uint32 u32;
+case SCV_I32:
+    int32 i32;
+case SCV_STATIC:
+    SCStatic ic;
+case SCV_OBJECT:
+    SCObject* obj;
+case SCV_SYMBOL:
+    SCSymbol sym;
+case SCV_BITSET:
+    uint64 bits;
+case SCV_STATUS:
+    SCStatus status;
+};
+
+enum SCObjectType
+{
+    // We have a few objects that represent non-stellar-specific concepts
+    // like general-purpose maps, vectors, strings, numbers, blobs, etc.
+
+    SCO_BOX = 0,
+    SCO_VEC = 1,
+    SCO_MAP = 2,
+    SCO_U64 = 3,
+    SCO_I64 = 4,
+    SCO_STRING = 5,
+    SCO_BINARY = 6,
+    SCO_BIGINT = 7,
+    SCO_BIGRAT = 8,
+
+    // Followed by a potentially much longer list of object types
+    // corresponding to any XDR ledger types users might want to manipulate
+    // directly. Separate type codes are required for composite types and
+    // each of their members, if it is reasonable for a user to want to hold
+    // a handle to the member separately from the composite type.
+
+    SCO_LEDGERKEY = 9,
+    SCO_OPERATION = 10,
+    SCO_OPERATION_RESULT = 11,
+    SCO_TRANSACTION = 12,
+    SCO_ASSET = 13,
+    SCO_PRICE = 14,
+    SCO_ACCOUNTID = 15
```

Comment:
I feel like we've had some discussion where we should remove all these values and remove string as well, and then reintroduce them as part of the interop work that will define them. Do you think we get anything from introducing them now?

Reply:
Eh, probably not. But we'll have to cut the SCO* vals out of the host when you next regenerate the XDR.

Reply:
It looks like we don't have a ton of code around those types, so maybe fine to drop?
<img width="410" alt="Screen Shot 2022-05-13 at 9 11 06 PM" src="https://user-images.githubusercontent.com/351529/168410259-07559104-a426-4ecf-9cef-64dc73cf4daf.png">
<img width="764" alt="Screen Shot 2022-05-13 at 9 11 18 PM" src="https://user-images.githubusercontent.com/351529/168410260-85b9ccdd-d88e-407f-a000-4e023873b62b.png">
There's some code for the String types, but it is pretty minimal.

Reply:
Deleting here presently. Pushed a branch of the env crate https://github.com/stellar/rs-stellar-contract-env/pull/65 with most of that deleted (kept obj_cmp and obj_{to,from}_{u64,i64} since they're fairly obvious and required for quite basic guest plumbing.

Reply:
I think it is also safe to assume vec + map too if you want to keep them. We already have them impl in the SDK too.